### PR TITLE
HyperRectangle and HyperCube refactor

### LINKED ISSUE #34: HyperCube and HyperRectangle are the same thing
See comments in #33. Specifically `HyperCube` allows each interval to be of different length, which goes against the hyper geometry definition. 

Since changing `HyperCube` itself is not enough work (:P), I propose switching from specifying intervals with start and end points in `HyperRectangle` to the approach in `HyperCube`. 

This is the proposed change:
**From:**
```
immutable HyperRectangle{N, T} <: GeometryPrimitive{N}
    minimum::Vec{N, T}
    maximum::Vec{N, T}
end
immutable HyperCube{N, T} <: GeometryPrimitive{N}
    origin::Vec{N, T}
    width::Vec{N, T}
end
```
**To:**
```
immutable HyperRectangle{N, T} <: GeometryPrimitive{N}
    origin::Vec{N, T}
    widths::Vec{N, T}
end
immutable HyperCube{N, T} <: GeometryPrimitive{N}
    origin::Vec{N, T}
    width::T
end
```

### PR DESCRIPTION
**Pull request by sjkelly**
Incorporating #33 and would close #34.

Quite a bit needs to be re-jiggered since the constructors in the tests are all now invalid, along with most of the algorithms.

### CODE REVIEW

**Comment by sjkelly** [comment]
@SimonDanisch do you have time review this?

File: src/hyperrectangles.jl [hunk 1]
```diff
@@ -1,7 +1,7 @@
-maximum(b::HyperRectangle) = b.maximum
-minimum(b::HyperRectangle) = b.minimum
+maximum(b::HyperRectangle) = b.origin + b.widths
+minimum(b::HyperRectangle) = b.origin
 length{T, N}(b::HyperRectangle{N, T}) = N
-width(b::HyperRectangle) = maximum(b) - minimum(b)
+widths{N,T}(b::HyperRectangle{N,T}) = b.widths
```
Comment (SimonDanisch) [inline src/hyperrectangles.jl]: We could start removing all field access... So we'd also need the function `origin`

File: src/hyperrectangles.jl [hunk 2]
```diff
@@ -53,10 +63,10 @@ end
 function call{N,T}(::Type{HyperRectangle{N,T}}, r::SimpleRectangle)
     if N > 2
         return HyperRectangle(Vec{N, T}(T(r.x), T(r.y), [zero(T) for i=1:N-2]...),
```
Comment (SimonDanisch) [inline src/hyperrectangles.jl]: maybe also remove field access?

File: src/hyperrectangles.jl [hunk 3]
```diff
@@ -124,12 +134,14 @@ function call{N1, N2, T1, T2}(t::Type{HyperRectangle{N1, T1}}, geometry::Array{P
          vmin = min(p, vmin)
          vmax = max(p, vmax)
     end
+    o = vmin
+    w = vmax - vmin
     if N1 > N2
-        return HyperRectangle{N1,T1}(Vec{N1,T1}(vmin..., [zero(T1) for i = 1:N1-N2]...),
-                                     Vec{N1,T1}(vmax..., [zero(T1) for i = 1:N1-N2]...))
+        return HyperRectangle{N1,T1}(Vec{N1,T1}(o..., [zero(T1) for i = 1:N1-N2]...),
```
Comment (SimonDanisch) [inline src/hyperrectangles.jl]: `Vec{N1,T1}(o, Vec{N1-N2, T1)...)` should be more efficient.

**Comment by SimonDanisch** [comment]
Thanks for doing this, looks great :)

File: src/hyperrectangles.jl (hunk 1 above)
Comment (sjkelly) [inline src/hyperrectangles.jl]: Good idea.

**Pull request merged**